I think I got a crash while creating a widget: 

```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.isoron.uhabits, PID: 1780
    java.lang.RuntimeException: Unable to bind to service org.isoron.uhabits.widgets.StackWidgetService@d5ed17e with Intent { dat=intent: cmp=org.isoron.uhabits/.widgets.StackWidgetService (has extras) }: java.lang.NumberFormatException: For input string: ""
        at android.app.ActivityThread.handleBindService(ActivityThread.java:4560)
        at android.app.ActivityThread.-$$Nest$mhandleBindService(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2167)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7898)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
     Caused by: java.lang.NumberFormatException: For input string: ""
        at java.lang.Long.parseLong(Long.java:746)
        at java.lang.Long.parseLong(Long.java:861)
        at org.isoron.platform.utils.StringUtils$Companion.splitLongs(StringUtils.kt:28)
        at org.isoron.uhabits.widgets.StackRemoteViewsFactory.<init>(StackWidgetService.kt:168)
        at org.isoron.uhabits.widgets.StackWidgetService.onGetViewFactory(StackWidgetService.kt:43)
        at android.widget.RemoteViewsService.onBind(RemoteViewsService.java:241)
        at android.app.ActivityThread.handleBindService(ActivityThread.java:4545)
        at android.app.ActivityThread.-$$Nest$mhandleBindService(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2167) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loopOnce(Looper.java:201) 
        at android.os.Looper.loop(Looper.java:288) 
        at android.app.ActivityThread.main(ActivityThread.java:7898) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936) 
W/System.err: java.lang.RuntimeException: Unable to bind to service org.isoron.uhabits.widgets.StackWidgetService@d5ed17e with Intent { dat=intent: cmp=org.isoron.uhabits/.widgets.StackWidgetService (has extras) }: java.lang.NumberFormatException: For input string: ""
W/System.err:     at android.app.ActivityThread.handleBindService(ActivityThread.java:4560)
W/System.err:     at android.app.ActivityThread.-$$Nest$mhandleBindService(Unknown Source:0)
W/System.err:     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2167)
W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:106)
W/System.err:     at android.os.Looper.loopOnce(Looper.java:201)
W/System.err:     at android.os.Looper.loop(Looper.java:288)
W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:7898)
W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
W/System.err:     at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
W/System.err: Caused by: java.lang.NumberFormatException: For input string: ""
W/System.err:     at java.lang.Long.parseLong(Long.java:746)
W/System.err:     at java.lang.Long.parseLong(Long.java:861)
W/System.err:     at org.isoron.platform.utils.StringUtils$Companion.splitLongs(StringUtils.kt:28)
W/System.err:     at org.isoron.uhabits.widgets.StackRemoteViewsFactory.<init>(StackWidgetService.kt:168)
W/System.err:     at org.isoron.uhabits.widgets.StackWidgetService.onGetViewFactory(StackWidgetService.kt:43)
W/System.err:     at android.widget.RemoteViewsService.onBind(RemoteViewsService.java:241)
W/System.err:     at android.app.ActivityThread.handleBindService(ActivityThread.java:4545)
W/System.err: 	... 9 more
I/Process: Sending signal. PID: 1780 SIG: 9
```



Not sure if it's just my emulator (which also crashed afterwards) or the app...

After rebooting it, the simple widgets seem to work, but the stacked widget tends to disappear for some reason, see video

https://user-images.githubusercontent.com/3812042/188500538-b7b0ffb5-846a-49b0-9ed3-f912d9a5e7db.mp4

--- a/uhabits-android/src/main/java/org/isoron/uhabits/widgets/StackWidgetService.kt
+++ b/uhabits-android/src/main/java/org/isoron/uhabits/widgets/StackWidgetService.kt
@@ -88,8 +88,26 @@ internal class StackRemoteViewsFactory(private val context: Context, intent: Int
     }
 
     override fun getViewAt(position: Int): RemoteViews? {
-        Log.i("StackRemoteViewsFactory", "getViewAt $position")
-        return if (0 <= position && position < remoteViews.size) remoteViews[position] else null
+        Log.i("StackRemoteViewsFactory", "getViewAt $position started")
+        if (position < 0 || position >= remoteViews.size) return null
+        val app = context.applicationContext as HabitsApplication
+        val prefs = app.component.preferences
+        val habitList = app.component.habitList
+        val options = AppWidgetManager.getInstance(context).getAppWidgetOptions(widgetId)
+        if (Looper.myLooper() == null) Looper.prepare()
+        val habits = habitIds.map { habitList.getById(it) ?: throw HabitNotFoundException() }
+        val h = habits[position]
+        val widget = constructWidget(h, prefs)
+        widget.setDimensions(getDimensionsFromOptions(context, options))
+        val landscapeViews = widget.landscapeRemoteViews
+        val portraitViews = widget.portraitRemoteViews
+        val factory = PendingIntentFactory(context, IntentFactory())
+        val intent = StackWidgetType.getIntentFillIn(factory, widgetType, h, habits, getToday())
+        landscapeViews.setOnClickFillInIntent(R.id.button, intent)
+        portraitViews.setOnClickFillInIntent(R.id.button, intent)
+        val remoteViews = RemoteViews(landscapeViews, portraitViews)
+        Log.i("StackRemoteViewsFactory", "getViewAt $position ended")
+        return remoteViews
     }
 
     private fun constructWidget(
@@ -134,28 +152,6 @@ internal class StackRemoteViewsFactory(private val context: Context, intent: Int
     }
 
     override fun onDataSetChanged() {
-        Log.i("StackRemoteViewsFactory", "onDataSetChanged started")
-        val app = context.applicationContext as HabitsApplication
-        val prefs = app.component.preferences
-        val habitList = app.component.habitList
-        val options = AppWidgetManager.getInstance(context).getAppWidgetOptions(widgetId)
-        val newRemoteViews = ArrayList<RemoteViews>()
-        if (Looper.myLooper() == null) Looper.prepare()
-        val habits = habitIds.map { habitList.getById(it) ?: throw HabitNotFoundException() }
-        for (h in habits) {
-            val widget = constructWidget(h, prefs)
-            widget.setDimensions(getDimensionsFromOptions(context, options))
-            val landscapeViews = widget.landscapeRemoteViews
-            val portraitViews = widget.portraitRemoteViews
-            val factory = PendingIntentFactory(context, IntentFactory())
-            val intent = StackWidgetType.getIntentFillIn(factory, widgetType, h, habits, getToday())
-            landscapeViews.setOnClickFillInIntent(R.id.button, intent)
-            portraitViews.setOnClickFillInIntent(R.id.button, intent)
-            newRemoteViews.add(RemoteViews(landscapeViews, portraitViews))
-            Log.i("StackRemoteViewsFactory", "onDataSetChanged constructed widget ${h.id}")
-        }
-        remoteViews = newRemoteViews
-        Log.i("StackRemoteViewsFactory", "onDataSetChanged ended")
     }
 
     init {

--- a/uhabits-android/src/main/java/org/isoron/uhabits/widgets/StackWidgetService.kt
+++ b/uhabits-android/src/main/java/org/isoron/uhabits/widgets/StackWidgetService.kt
@@ -57,7 +57,6 @@ internal class StackRemoteViewsFactory(private val context: Context, intent: Int
     )
     private val habitIds: LongArray
     private val widgetType: StackWidgetType
-    private var remoteViews = ArrayList<RemoteViews>()
     override fun onCreate() {}
     override fun onDestroy() {}
     override fun getCount(): Int {
@@ -89,7 +88,7 @@ internal class StackRemoteViewsFactory(private val context: Context, intent: Int
 
     override fun getViewAt(position: Int): RemoteViews? {
         Log.i("StackRemoteViewsFactory", "getViewAt $position started")
-        if (position < 0 || position >= remoteViews.size) return null
+        if (position < 0 || position >= habitIds.size) return null
         val app = context.applicationContext as HabitsApplication
         val prefs = app.component.preferences
         val habitList = app.component.habitList

--- a/uhabits-core/src/commonMain/kotlin/org/isoron/platform/utils/StringUtils.kt
+++ b/uhabits-core/src/commonMain/kotlin/org/isoron/platform/utils/StringUtils.kt
@@ -25,6 +25,12 @@ class StringUtils {
 
         fun joinLongs(values: LongArray): String = values.joinToString(separator = ",")
 
-        fun splitLongs(str: String): LongArray = str.split(",").map { it.toLong() }.toLongArray()
+        fun splitLongs(str: String): LongArray {
+            return try {
+                str.split(",").map { it.toLong() }.toLongArray()
+            } catch (e: NumberFormatException) {
+                LongArray(0)
+            }
+        }
     }
 }

